addons/account/models/account_tax.py

```diff
@@ -552,90 +552,688 @@ def onchange_price_include(self):
-            # of the quantity and the sign of the price_unit
-            # Amount is the fixed price for the tax, it can be negative
-            # Base amount included the sign of the quantity and the sign of the unit price and when
-            # a product is returned, it can be done either by changing the sign of quantity or by changing the
-            # sign of the price unit.
-            # When the price unit is equal to 0, the sign of the quantity is absorbed in base_amount then
-            # a "else" case is needed.
-            if base_amount:
-                return math.copysign(quantity, base_amount) * self.amount * abs(fixed_multiplicator)
-            else:
-                return quantity * self.amount * abs(fixed_multiplicator)
-
-        price_include = self._context.get('force_price_include', self.price_include)
-
-        # base * (1 + tax_amount) = new_base
-        if self.amount_type == 'percent' and not price_include:
-            return base_amount * self.amount / 100
-        # <=> new_base = base / (1 + tax_amount)
-        if self.amount_type == 'percent' and price_include:
-            return base_amount - (base_amount / (1 + self.amount / 100))
-        # base / (1 - tax_amount) = new_base
-        if self.amount_type == 'division' and not price_include:
-            return base_amount / (1 - self.amount / 100) - base_amount if (1 - self.amount / 100) else 0.0
-        # <=> new_base * (1 - tax_amount) = base
-        if self.amount_type == 'division' and price_include:
-            return base_amount - (base_amount * (self.amount / 100))
-        # default value for custom amount_type
-        return 0.0
-
-    def json_friendly_compute_all(self, price_unit, currency_id=None, quantity=1.0, product_id=None, partner_id=None, is_refund=False, include_caba_tags=False):
-        """ Called by the reconciliation to compute taxes on writeoff during bank reconciliation
+    def _convert_to_dict_for_taxes_computation(self):
+        """ Convert the current taxes to a list of dict.
+        Since a lot of method here are copy-pasted in account_tax.js, we want to keep exactly the same method in both
+        python and javascript.
+
+        :return: A list of dictionaries, each one representing the taxes.
+        """
+        return [tax._prepare_dict_for_taxes_computation() for tax in self]
+
+    # -------------------------------------------------------------------------
+    # HELPERS IN BOTH PYTHON/JAVASCRIPT (account_tax.js / account_tax.py)
+
+    # PREPARE TAXES COMPUTATION
+    # -------------------------------------------------------------------------
+
+    @api.model
+    def _prepare_taxes_batches(self, tax_values_list):
+        """ Group the taxes passed as parameter by nature because some taxes must be computed all together
+        like price-included percent or division taxes.
+
+        [!] Mirror of the same method in account_tax.js.
+        PLZ KEEP BOTH METHODS CONSISTENT WITH EACH OTHERS.
```

Comment:
@qdp-odoo Tests are also duplicated and are executed both in python and js. No need a tour

Reply:
well, instead of a test verifying the computation in python and a test verifying the computation in JS, we could have a test verifying the computation in python, and a test verifying the JS gives the same results. The problem we currently might have is someone fixing/modifying a behavior in python and forgetting to mimic that in JS (or the other way around)... and that could circumvent it I think.

+ the test code is duplicated besides the computation code.

So I definitively think this is a potential improvement, for a next task